### PR TITLE
fix for issue #128 Use Apache solr for search 

### PR DESCRIPTION
**Pull request by rystills**
sunspot_rails and sunspot_solr have been added to the gemfile, and the Course search SQL query has been replaced with a simple Sunspot search. The accuracy of search results still needs to be improved, and  weighting needs to be adjusted (eg. so titles are weighted more strongly than descriptions).

### CODE REVIEW

**Comment by rystills** [comment]
This should now be fully functional with partial as well as complete matches, and has been weighted appropriately, as well as tweaked a bit to match the original SQL query.

File: app/models/course.rb [hunk 1]
```diff
@@ -3,32 +3,26 @@ class Course < ActiveRecord::Base
...
-      LIMIT 25;
-    SQL
-    courses = find_by_sql(query).uniq
+    courses = Sunspot.search(Course) do
+      fulltext params do
+	    fields(:description, :number => 10, :name => 5.0) #weight results most strongly by number, then name, and finally description
+	    phrase_fields :name => 3.0 #weight titles that contain a phrase of search terms more strongly
+	  end
+	  paginate :page => 1, :per_page => 25 #place a maximum of 25 results on each page (for now we just display page 1)
+    end.results.uniq #store unique search results in courses in one step
```
Comment (Bad-Science) [inline app/models/course.rb]: Can we make solr handle only returning unique results? uniq in ruby is relatively slow.

File: config/sunspot.yml [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    path: /solr/development
+
+test: &TEST
+  solr:
+    hostname: localhost
+    port: 8981
+    log_level: WARNING
+    path: /solr/test
+
+cucumber:
```
Comment (Bad-Science) [inline config/sunspot.yml]: cucumber is not actually an environment: cucumber tests run in the test environment. was this line generated or added after?

File: config/sunspot.yml [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    port: 8983
+    log_level: WARNING
+    path: /solr/production
+    # read_timeout: 2
+    # open_timeout: 0.5
+
+development:
+  solr:
+    hostname: localhost
+    port: 8982
```
Comment (Bad-Science) [inline config/sunspot.yml]: i'd prefer to run it on one port, we'll see if we can make that work.

File: Gemfile [hunk 4]
```diff
@@ -58,3 +58,8 @@ group :development do
   gem 'web-console', '~> 2.0'
   gem 'spring'
 end
+
+gem 'tzinfo-data' #timezone gem for hosting on windows
```
Comment (Bad-Science) [inline Gemfile]: note: we'll want to remove this when we merge this work with docker

File: config/sunspot.yml (hunk 2 above)
Comment (rystills) [inline config/sunspot.yml]: Good catch; this was part of a solution I had tried for getting stubbing to work in cucumber, but turned out not to be needed. I'll go ahead and revert it.

File: app/models/course.rb (hunk 1 above)
Comment (rystills) [inline app/models/course.rb]: I noticed that we were calling uniq after the original SQL query, so I included it here as well, but I'm not entirely sure what constitutes unique results. Does our database have some identical duplicate class data in it (same exact name, number of credits, description, etc..)? The closest thing Sunspot has may be grouping/field collapsing http://wiki.apache.org/solr/FieldCollapsing

**Comment by Bad-Science** [comment]
No there are no identical classes, this was only needed due to the way I
formed the query. It should be able to be removed.

**Comment by copperwater** [comment]
We want to merge this into docker, not staging, and then later have our big merge of docker into staging.
@rystills please change the base branch.

File: nginx/nginx.conf [hunk 5]
```diff
@@ -0,0 +1,28 @@
+events {
```
Comment (Bad-Science) [inline nginx/nginx.conf]: please remove this file

File: nginx/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,3 @@
+FROM nginx
```
Comment (Bad-Science) [inline nginx/Dockerfile]: please remove this file

**Comment by rystills** [comment]
Closing - this PR has been moved to #194 - 'Integrate Apache solr search with docker' as solr will be deployed with docker going forward.

File: nginx/nginx.conf (hunk 5 above)
Comment (rystills) [inline nginx/nginx.conf]: Removed in PR #194

**Pull request closed without merging**